storage/src/test/java/org/apache/kafka/server/log/remote/storage/LocalTieredStorage.java

```diff
@@ -0,0 +1,578 @@
+                } catch (IOException e) {
+                    throw new RuntimeException(format("Not able to create the storage directory '%s'", storageDirectory.getAbsolutePath()), e);
+                }
+            } else {
+                logger.warn("Remote storage with ID [{}] already exists on the file system. Any data already " +
+                        "in the remote storage will not be deleted and may result in an inconsistent state and/or " +
+                        "provide stale data.", storageDir);
+            }
+        }
+
+        logger.info("Created local tiered storage manager [{}]:[{}]", brokerId, storageDirectory.getName());
+    }
+
+    @Override
+    public void copyLogSegmentData(final RemoteLogSegmentMetadata metadata, final LogSegmentData data)
+            throws RemoteStorageException {
+        Callable<Void> callable = () -> {
+            final RemoteLogSegmentId id = metadata.remoteLogSegmentId();
+            final LocalTieredStorageEvent.Builder eventBuilder = newEventBuilder(OFFLOAD_SEGMENT, id);
+            RemoteLogSegmentFileset fileset = null;
+
+            try {
+                fileset = openFileset(storageDirectory, id);
+
+                logger.info("Offloading log segment for {} from segment={}", id.topicIdPartition(), data.logSegment());
+
+                fileset.copy(transferer, data);
+
+                storageListeners.onStorageEvent(eventBuilder.withFileset(fileset).build());
+
+            } catch (final Exception e) {
+                //
+                // Keep the storage in a consistent state, i.e. a segment stored should always have with its
+                // associated offset and time indexes stored as well. Here, delete any file which was copied
+                // before the exception was hit. The exception is re-thrown as no remediation is expected in
+                // the current scope.
+                //
+                if (fileset != null) {
+                    fileset.delete();
+                }
+
+                storageListeners.onStorageEvent(eventBuilder.withException(e).build());
+                throw e;
+            }
+
+            return null;
+        };
+
+        wrap(callable);
+    }
+
+    @Override
+    public InputStream fetchLogSegment(final RemoteLogSegmentMetadata metadata,
+                                       final int startPos) throws RemoteStorageException {
+        return fetchLogSegment(metadata, startPos, Integer.MAX_VALUE);
+    }
+
+    @Override
+    public InputStream fetchLogSegment(final RemoteLogSegmentMetadata metadata,
+                                       final int startPos,
+                                       final int endPos) throws RemoteStorageException {
+        checkArgument(startPos >= 0, "Start position must be positive", startPos);
+        checkArgument(endPos >= startPos,
+                "End position cannot be less than startPosition", startPos, endPos);
+
```

Comment:
maybe also validate `metadata.segmentSizeInBytes >= endPos`?

Reply:
`endPos` is currently unused in the LocalRemoteStorage. We can refactor it when we are using the value.

Reply:
Updated.